src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -1917,19 +1922,41 @@ private void EmitArrayLength(BoundArrayLength expression, bool used)
             EmitPopIfUnused(used);
         }
 
-        private void EmitArrayCreationExpression(BoundArrayCreation expression, bool used)
+        private void EmitUninitializedArrayCreation(BoundArrayCreation expression, SyntaxNode syntax, MethodSymbol allocUninitialized)
         {
-            var arrayType = (ArrayTypeSymbol)expression.Type;
+            var arrLen = ConstantValue.Create(expression.InitializerOpt.Initializers.Length);
+            var pinned = ConstantValue.Create(false);
 
-            EmitArrayIndices(expression.Bounds);
+            var arg1 = new BoundLiteral(syntax, arrLen, _module.Compilation.GetSpecialType(SpecialType.System_Int32));
+            var arg2 = new BoundLiteral(syntax, pinned, _module.Compilation.GetSpecialType(SpecialType.System_Boolean));
 
+            var arguments = ImmutableArray.Create<BoundExpression>(arg1, arg2);
+
+            EmitStaticCall(allocUninitialized, null, arguments, UseKind.UsedAsValue, syntax,
+                                ImmutableArray.Create(RefKind.None, RefKind.None));
+        }
+
+        private void EmitArrayCreationExpression(BoundArrayCreation expression, bool used)
+        {
+            var arrayType = (ArrayTypeSymbol)expression.Type;
             if (arrayType.IsSZArray)
             {
-                _builder.EmitOpCode(ILOpCode.Newarr);
-                EmitSymbolToken(arrayType.ElementType, expression.Syntax);
+                var allocUninitialized = _module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_GC__AllocateUninitializedArray_T);
+                if (expression.InitializerOpt != null && allocUninitialized is MethodSymbol { } alloc)
+                {
+                    var constructed = alloc.Construct(ImmutableArray.Create(arrayType.ElementType));
+                    EmitUninitializedArrayCreation(expression, expression.Syntax, constructed);
```

Comment:
>expression [](http://example.com/codeflow?start=51&length=10)

nit: the only thing we need from `expression` parameter is the length of initializers. Consider moving `expression.InitializerOpt.Initializers.Length` to caller and using an `int` parameter